modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/IOConnection.kt

```diff
@@ -0,0 +1,112 @@
+package arrow.effects.internal
+
+import arrow.effects.ForIO
+import arrow.effects.IO
+import arrow.effects.fix
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Represents a composite of functions (meant for cancellation) that are stacked. cancel() is idempotent,
+ * and all methods are thread-safe & atomic.
+ */
+sealed class IOConnection {
+
+  /**
+   * Cancels the unit of work represented by this reference.
+   *
+   * Guaranteed idempotency - calling it multiple times should have the same side-effect as calling it only
+   * once. Implementations of this method should also be thread-safe.
+   */
+  abstract fun cancel(): CancelToken<ForIO>
```

Comment:
All the `CancelToken` in the PR are fixed to `ForIO`. Doesn't this imply that there should be other fixes for other datatypes?

Reply:
Yep. We need to support Cancellation over any `F` but that's part of [ConcurrentEffect](https://github.com/typelevel/cats-effect/blob/76d354153926b30217cc1aed9d1b181c7ea20d32/site/src/main/tut/typeclasses/concurrent-effect.md) and [Concurrent](https://github.com/typelevel/cats-effect/blob/77931fd2fc6d7d2891e321cd59689a47b2e13a56/site/src/main/tut/typeclasses/concurrent.md). That will be done when we work on those types completely (we just have `ConcurrentEffect` on very early steps for now).